### PR TITLE
[instancer] CFF2 instancing

### PR DESCRIPTION
**Pull request by behdad**
WIP. Almost there.

### CODE REVIEW

**Comment by behdad** [comment]
I put the blends into the VarStore, instantiate that, and rebuild the blends from it.

Since we have code to instantiate VarStore, it works nicely.

**Comment by behdad** [comment]
hmtx/vmtx need to be updated as well. The side-bearings are very hard to update (and unused for CFF).

**Comment by behdad** [comment]
Did the advance widths.

**Comment by behdad** [comment]
Oh, and pruning the VarStore; and dropping vsindex when not needed and VarStore when not needed.

**Comment by behdad** [comment]
> Oh, and pruning the VarStore; and dropping vsindex when not needed and VarStore when not needed.

Done.

**Comment by behdad** [comment]
> > * The font private program needs to be instantiated too. Donno how hard that one is.

I know how to do it. But can't find any fonts using those. So, for now, I left it unimplemented error.

I think this PR is ready. Volunteers for tests needed.

**Comment by behdad** [comment]
We might also want to downgrade to CFF if static instancing.

**Comment by behdad** [comment]
> > > * The font private program needs to be instantiated too. Donno how hard that one is.
> 
> I know how to do it. But can't find any fonts using those. So, for now, I left it unimplemented error.

This done now as well. I manually constructed a font.

**Comment by behdad** [comment]
With test now.

**Comment by behdad** [comment]
Looks like we only have code to upgrade CFF to CFF2, but not the other way around. Shouldn't be hard to write.

**Comment by behdad** [comment]
It occurred to me that if we downgrade to CFF, we also have to remove overlaps. Is that desirable?

**Comment by punchcutter** [comment]
Thanks for working on this. I've been waiting a long time to be able to do this with fontTools. Definitely would be nice to optionally remove overlaps with a flag.

**Comment by behdad** [comment]
> Looks like we only have code to upgrade CFF to CFF2, but not the other way around. Shouldn't be hard to write.

I implemented a very rudimentary CFF2ToCFF. Doesn't remove overlaps, but seems to do enough that ots passes and both svgPathPen and hb-view show text. My confidence is rather low though, I'm sure there are incomplete corners.

**Comment by behdad** [comment]
> > 
> 
> I implemented a very rudimentary CFF2ToCFF. Doesn't remove overlaps, but seems to do enough that ots passes and both svgPathPen and hb-view show text. My confidence is rather low though, I'm sure there are incomplete corners.

While writing that I found that CFFToCFF2 is missing a lot actually. It doesn't add endchar and return ops to charstrings for example.

**Comment by behdad** [comment]
> We have two convertCFFToCFF2 #1835

Thanks. Very timely. I'm working on it.

**Comment by behdad** [comment]
> > We have two convertCFFToCFF2 #1835
> 
> Thanks. Very timely. I'm working on it.

Humm. M'okay. It's a bit more work. I'll do it in a separate PR. Can I get a review on this one?

File: Lib/fontTools/cffLib/CFFToCFF2.py [hunk 1]
```diff
@@ -101,9 +102,66 @@ def _convertCFFToCFF2(cff, otFont):
...
+        program = cs.program
+        if fdIndex == None:
+            fdIndex = 0
+
+        # Intentionally use None for nominalWidthX, such that any
+        # CharString that has an explicit width encoded will throw.
+        extractor = T2WidthExtractor(localSubrs[fdIndex], globalSubrs, None, 0)
+        try:
+            extractor.execute(cs)
+        except TypeError:
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/CFFToCFF2.py]: TypeError seems a bit too broad here, I wonder if the T2WidthExtractor could throw some more specific exception that you can catch here?

Comment (behdad) [reply Lib/fontTools/cffLib/CFFToCFF2.py]: Yeah I don't like it either. I'll make a custom object that throws a custom exception on addition. No need to change the `T2WidthExtractor`.

Comment (behdad) [reply Lib/fontTools/cffLib/CFFToCFF2.py]: The merger uses a custom object as defaultWidth instead and detects if that was returned as width. But this way using exceptions stops as soon as the question is answered so I went with this model this time.

File: Lib/fontTools/cffLib/CFFToCFF2.py [hunk 2]
```diff
@@ -101,9 +102,66 @@ def _convertCFFToCFF2(cff, otFont):
...
+            # Program has explicit width. We want to drop it, but can't
+            # just pop the first number since it may be a subroutine call.
+            # Instead, when seeing that, we embed the subroutine and recurse.
+            # This has the problem that some subroutines might become unused.
+            # We don't currently prune those. Subset module has code for this
+            # kind of stuff, possibly plug it in here if pruning becomes needed.
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/CFFToCFF2.py]: +1 on moving the subset's subrs pruning to a place where can be reused here

File: Lib/fontTools/cffLib/CFFToCFF2.py (hunk 1 above)
Comment (anthrotype) [reply Lib/fontTools/cffLib/CFFToCFF2.py]: i'm simply concerned that a TypeError may trigger inadvertently for unrelated bugs, activating the subrs-embeddindg-width-popping logic below

File: Lib/fontTools/cffLib/CFFToCFF2.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    # TODO(behdad): What does the following comment even mean? Both CFF and CFF2
+    # use the same T2Charstring class.
+
+    # At this point, the Subrs and Charstrings are all still T2Charstring class
+    # easiest to fix this by compiling, then decompiling again
+    file = BytesIO()
+    cff.compile(file, otFont, isCFF2=True)
+    file.seek(0)
+    cff.decompile(file, otFont, isCFF2=True)
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/CFFToCFF2.py:177]: could it be that this block is no longer needed after you added the T2WidthExtractor thing?

File: Lib/fontTools/cffLib/CFF2ToCFF.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
+"""CFF to CFF2 converter."""
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/CFF2ToCFF.py:1]:
```suggestion
"""CFF2 to CFF converter."""
```

File: Lib/fontTools/cffLib/CFF2ToCFF.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+
+log = logging.getLogger("fontTools.cffLib")
+
+
+def _convertCFF2ToCFF(cff, otFont):
+    """Converts this object from CFF2 format to CFF format. This conversion
+    is done 'in-place'. The conversion cannot be reversed.
+
+    The CFF2 font cannot be variable. (TODO Accept those and convert to the
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/CFF2ToCFF.py:21]: is the input CFF2 also assumed to _not_ be already subroutinized? Or it doesn't matter in this case?

Comment (anthrotype) [reply Lib/fontTools/cffLib/CFF2ToCFF.py:21]: oh I see you do seem to handle GlobalSubrs and Private.Subrs in here. I was simply wondering if inserting the optimal width at index 0 can be done unconditionally even in presence of subroutines. Ignore me if it doesn't make any sense

File: Lib/fontTools/varLib/instancer/__init__.py [hunk 6]
```diff
@@ -765,22 +1024,52 @@ def _remapVarIdxMap(table, attrName, varIndexMapping, glyphOrder):
...
+            metrics = metricsTable.metrics
+            for glyphName, (advanceWidth, sb) in metrics.items():
+                if advMapping:
+                    varIdx = advMapping.mapping[glyphName]
+                else:
+                    varIdx = varfont.getGlyphID(glyphName)
+                metrics[glyphName] = (advanceWidth + round(defaultDeltas[varIdx]), sb)
+
+            # TODO if tableTag == "VVAR" and getattr(vhvar, tableFields.vOrigMapping),
+            # update VORG table as well
```
Comment (anthrotype) [inline Lib/fontTools/varLib/instancer/__init__.py:1066]: how about raising some error if VORG is present? a user might not notice this is not implemented

**Review by anthrotype — APPROVED** [review]

File: Lib/fontTools/cffLib/CFF2ToCFF.py (hunk 5 above)
Comment (behdad) [reply Lib/fontTools/cffLib/CFF2ToCFF.py:21]: Yeah there's no problem with subroutines. I first compute optimal default and nominal widths from the hmtx table data, then just insert at index 0 the width for those glyphs that need it. It works with subroutines as well.

File: Lib/fontTools/cffLib/CFFToCFF2.py (hunk 3 above)
Comment (behdad) [reply Lib/fontTools/cffLib/CFFToCFF2.py:177]: Humm. I'm not sure. I'll check. Maybe. But even with the block I was getting ots fail re the width without explicitly implementing it.

It talks about T2Charstring class... @readroberts maybe is still around?

I'll go ahead and delete it I think.

Comment (behdad) [reply Lib/fontTools/cffLib/CFFToCFF2.py]: I added a comment re what I think it's doing. It's fine for now. If it's slow we can look into removing and implement what I suggest in the comment.

File: Lib/fontTools/varLib/instancer/__init__.py (hunk 6 above)
Comment (behdad) [reply Lib/fontTools/varLib/instancer/__init__.py]: I logged a warning.

File: Lib/fontTools/cffLib/CFFToCFF2.py (hunk 2 above)
Comment (behdad) [reply Lib/fontTools/cffLib/CFFToCFF2.py]: https://github.com/fonttools/fonttools/issues/3511

**Pull request merged**